src/BidirectedGraph.cpp

```diff
@@ -114,14 +117,102 @@ void BidirectedGraph::mod_BFS(uint64_t id){
         }
         reachable++;
     }
-}
\ No newline at end of file
+}
+
+/// Method to check if a node exists by ID
+bool BidirectedGraph::has_node(nid_t nodeid) const {
+    return edges.find(nodeid) != edges.end();
+}
+
+/// Look up the handle for the node with the given ID in the given orientation
+handle_t BidirectedGraph::get_handle(const nid_t& node_id, bool is_reverse) const {
+    return number_bool_packing::pack(node_id, is_reverse);
+}
+
+/// Get the ID from a handle
+nid_t BidirectedGraph::get_id(const handle_t& handle) const {
+    return number_bool_packing::unpack_number(handle);
+}
+
+/// Get the orientation of a handle
+bool BidirectedGraph::get_is_reverse(const handle_t& handle) const {
+    return number_bool_packing::unpack_bit(handle);
+}
+
+/// Invert the orientation of a handle (potentially without getting its ID)
+handle_t BidirectedGraph::flip(const handle_t& handle) const {
+    return number_bool_packing::toggle_bit(handle);
+}
+
+/// Get the length of a node
+size_t BidirectedGraph::get_length(const handle_t& handle) const {
+    return 0;
+}
+
+/// Get the sequence of a node, presented in the handle's local forward
+/// orientation.
+std::string BidirectedGraph::get_sequence(const handle_t& handle) const {
+    return "";
+}
+
+/// Return the number of nodes in the graph
+size_t BidirectedGraph::get_node_count() const {
+    return edges.size();
+}
+
+/// Return the smallest ID in the graph, or some smaller number if the
+/// smallest ID is unavailable. Return value is unspecified if the graph is empty.
+nid_t BidirectedGraph::min_node_id() const {
+    // Should work with default comparator, otherwise use
+    edge_map::const_iterator res = min_element(edges.begin(), edges.end(),
+                                        [](const edge_map_pair& node1, const edge_map_pair& node2) {
+                                            return node1.first < node2.first;
+                                        });
+    if (res != edges.end()) {
+        return res->first;
+    }
+    return 0;
+}
+
+/// Return the largest ID in the graph, or some larger number if the
+/// largest ID is unavailable. Return value is unspecified if the graph is empty.
+nid_t BidirectedGraph::max_node_id() const {
+    edge_map::const_iterator res = max_element(edges.begin(), edges.end(),
+                                        [](const edge_map_pair& node1, const edge_map_pair& node2) {
+                                            return node1.first < node2.first;
+                                        });
+    if (res != edges.end()) {
+        return res->first;
+    }
+    return 0;
+}
+
+bool BidirectedGraph::follow_edges_impl(const handle_t& handle, bool go_left, const std::function<bool(const handle_t&)>& iteratee) const {
+    nid_t node_id = get_id(handle);
+    for (const BidirectedEdge& node_edge : edges.at(node_id)) {
+        if (node_edge.from_left == go_left && !iteratee(get_handle(node_edge.id2, go_left))) {
```

Comment:
When we say "go left", that's "left relative to the orientation of the handle". That means you should be finding edges on the right side of the node if `go_left && get_is_reverse(handle)` or if `!go_left && !get_is_reverse(handle)`. Otherwise you'll want to find edges on the left side of the node. Also note that the handle you execute on for `go_left && get_is_reverse(handle)` will be reverse of the handle you execute on for `!go_left && !get_is_reverse(handle)`.

Also, make sure that you only call the iteratee on the edge targets that pass this filter. Right now you have the iteratee in the if condition, so it will actually execute on every target (even if the edge is on the wrong side of the current node).

Reply:
I just want to double check. If `go_left = true`, the I'll have to find the edges in the opposite direction of the handle's orientation?

Also correct me if this is wrong, but if the first part of the if statement doesn't evaluate to true won't C++ not evaluate the latter half? So in this case if the condition was `node_edge.from_left == go_left && !iteratee(get_handle(node_edge.id2, go_left)))` it'll only evaluate the second half if `node_edge.from_left == go_left`. Shouldn't be equivalent to 
```
if (node_edge.from_left == go_left) {                     // Check if edge's side is correct
    if (!iteratee(get_handle(node_edge.id2, go_left)))) { // Early termination check
        return false;                                     // If there is an early termination, return False
    }
}
```

Reply:
Based on my understanding I edited the code such that
```
bool BidirectedGraph::follow_edges_impl(const handle_t& handle, bool go_left, const std::function<bool(const handle_t&)>& iteratee) const {
    nid_t node_id    = get_id(handle);
    bool  is_reverse = get_is_reverse(handle);
    bool  on_left    = (go_left && !is_reverse) || (!go_left && is_reverse);
    for (const BidirectedEdge& node_edge : edges.at(node_id)) {
        if (node_edge.from_left != on_left && !iteratee(get_handle(node_id, is_reverse))) {
            return false;
        }
    }
    return true;
}
```